controlplane/kubeadm/controllers/controller.go

```diff
@@ -320,3 +320,37 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
 
 	return nil
 }
+
+// reconcileHealth performs health checks for control plane components and etcd
+// It removes any etcd members that do not have a corresponding node.
+func (r *KubeadmControlPlaneReconciler) reconcileHealth(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane, controlPlane *internal.ControlPlane) error {
+	logger := controlPlane.Logger()
+
+	// Do a health check of the Control Plane components
+	if err := r.managementCluster.TargetClusterControlPlaneIsHealthy(ctx, util.ObjectKey(cluster), kcp.Name); err != nil {
+		logger.V(2).Info("Waiting for control plane to pass control plane health check to continue reconciliation", "cause", err)
+		r.recorder.Eventf(kcp, corev1.EventTypeWarning, "ControlPlaneUnhealthy",
+			"Waiting for control plane to pass control plane health check to continue reconciliation: %v", err)
+		return &capierrors.RequeueAfterError{RequeueAfter: healthCheckFailedRequeueAfter}
+	}
+
+	// Ensure etcd is healthy
+	if err := r.managementCluster.TargetClusterEtcdIsHealthy(ctx, util.ObjectKey(cluster), kcp.Name); err != nil {
+		// If there are any etcd members that do not have corresponding nodes, remove them from etcd and from the kubeadm configmap.
+		// This will solve issues related to manual control-plane machine deletion.
+		workloadCluster, err := r.managementCluster.GetWorkloadCluster(ctx, util.ObjectKey(cluster))
+		if err != nil {
+			return err
+		}
+		if err := workloadCluster.ReconcileEtcdMembers(ctx); err != nil {
```

Comment:
I think it makes sense if we also fix KubeadmConfigMap there. If we reconcile KubeadmConfigMap in ControlPlane reconcile as you suggested above, it won't help as without fixing KubeadmConfigMap, this fix is not useful so we might requeue as well.

Reply:
I wouldn't necessarily expect there to be issues if we handled the reconciliation of etcd membership as part of the etcd health check and the reconciliation of the kubeadmconfigmap as part of the "control plane" health check, but it would likely introduce a small window where there could be eventual consistency drift between the two.

I'm good with sticking with the current approach for now.